File: design-patterns/JavaDesignPatterns/src/com/bigboxcode/designpattern/adapter/api/FileAdapter.java
// FileAdapter.java

package com.bigboxcode.designpattern.adapter.api;

public class FileAdapter implements Api {

    private final File file;

    public FileAdapter(File file) {
        this.file = file;
    }

    @Override
    public String fetchData() {
        // May perform additional operation or processing
        // before or after data is fetched

        return file.readFile();
    }

    @Override
    public void sendData(String data) {
        // May perform additional operation or processing
        // before or after data is written to file

        file.writeFile(data);
    }
}

File: design-patterns/go/adapter/file_api/file_adapter.go
// file_adapter.go

package main

type FileAdapter struct {
	file File
}

func NewFileAdapter(file File) (fileAdapter *FileAdapter) {
	fileAdapter = &FileAdapter{}
	fileAdapter.file = file
	return
}

func (fileAdapter *FileAdapter) FetchData() string {
	// May perform additional operation or processing
	// before or after data is fetched

	return fileAdapter.file.ReadFile()
}

func (fileAdapter *FileAdapter) SendData(data string) {
	// May perform additional operation or processing
	// before or after data is written to file

	fileAdapter.file.WriteFile(data)
}
